Add output format option for GitHub annotations

saibotk opened the pull request:
This PR implements a basic format option to enable output that is readable by the GitHub actions parser 
(see https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#setting-a-warning-message). 

This is particularly useful, when using glualint in a GitHub action workflow, to allow GitHub to parse the result and add annotations to the review.

The config option is backwards compatible and will by default still use the regular output format.

As this is my first time ever reading / writing Haskell, any feedback would be appreciated. I tried to keep it as simple as I could, but if there is anything, you would want to be changed / improved, let me know.

PS: Happy new year :rocket: :fireworks:

FPtje commented on src/GLuaFixer/LintMessage.hs:
```diff
@@ -2,17 +2,39 @@ module GLuaFixer.LintMessage where
+
+formatLintMessage :: LintMessage -> String -> String
+formatLintMessage lintMsg format = do
+  let (rg, msg, file, level) = getLintMessageData lintMsg
```
Instead of `getLintMessageData` you can do a case expression on `lintmsg`. It would look something like this:

```haskell
case lintMsg of
  LintError rg msg file -> ...
  LintWarning rg msg file -> ...
```

I think that case expression would fit best in `formatLintMessageGithub` and `FormatLintMessageDefault`

FPtje commented on src/GLuaFixer/LintSettings.hs:
```diff
@@ -37,6 +37,8 @@ data LintSettings =
     , prettyprint_rejectInvalidCode  :: !Bool
     , prettyprint_indentation        :: !String
+
+    , log_format                     :: !String
```
Since there are only two possible formats (at this time) I think `String` is not the ideal data type here. For such cases, it's better to create a custom data type:

```haskell
data LogFormat = StandardLogFormat | GithubLogFormat 
```

You'd then need a `FromJSON` instance to parse it from the `LintSettings`. Though I'm thinking the lint settings should perhaps be more for use with glualint in an editor. Wouldn't it be better to make the log format a command line parameter? That way the GitHub action could use that parameter to the the logs in the right format, while the linter in the scripter's favourite editor is unaffected.

saibotk replied:
Yes that was what I initially tried, but then I saw, that glualint usually goes in to different "branches" of code via the parameters and uses most of them more or less in a subcommand manner. That's why I added it to the config options, and also because I did not really know how to refactor the parameter logic to be fancy enough to support such a parameter :sweat_smile: 

I also wanted to add an `auto` mode, which tries to figure out which mode to use based on Environment variables (to automatically recognize running in a GitHub action / workflow), but i did not really get that far with using them in my code, as it always had some issues i do not yet understand, as i literally started with Haskell yesterday. So it would be cool to get some feedback on whether you think an `auto` mode would be useful and maybe some tips on how to check if an ENV var is set (as i do not care about its value) and I did not understand on how to evaluate just that (generally on how to check if it is not `Nothing`).

FPtje reviewed (changes requested):
This is a great idea! Using glualint for GitHub actions is a great idea, and it could definitely improve the quality of pull requests.

There are some comments and suggestions. Thanks for taking the effort to contribute to a project written in a language you've never used before! I'll be glad to help out if you get stuck somewhere :+1:

saibotk commented:
I refactored it to use a custom data type instead and it is now properly parsed from the JSON object. 
So we only need to decide on whether to keep it as an option in the config.json or if we would want to move it to the command line arguments (which I would need help with). 

And deciding on implementing an `auto` mode as explained in the comment above.

PS: This is what it could look like (The output of another tool, that I adjusted to work with the annotations :) )
![image](https://user-images.githubusercontent.com/10776964/103491845-3bdd9300-4e27-11eb-8c64-0ef6bbfe139c.png)

FPtje commented on src/GLuaFixer/LintMessage.hs:
```diff
@@ -1,18 +1,60 @@
 
 -- | Represents lint messages
 data LintMessage
   = LintError { err_rg :: Region, err_msg :: String, err_file :: String }
```
This is actually an early mistake of mine. The two constructors duplicate the message, region and file. It would have been better to have done the following:

```haskell
data Severity = Warning | Error

data LintMessage = LintMessage {lintmsg_severity :: Severity, lintmsg_msg :: String, lintmsg_file :: String}
```

That would make the formatting code prettier as well. But alas, back in the day this was one of my first Haskell projects, so I had much to learn myself 😉

FPtje commented on src/GLuaFixer/LintMessage.hs:
```diff
@@ -1,18 +1,60 @@
+
+formatLintMessageGithub :: LintMessage -> String
+formatLintMessageGithub lintMsg = do
+  let (rg, msg, file, level) = case lintMsg of
```
instead of the separate let binding, you can do `rg@(Region start _)`

FPtje reviewed (approved):
I realise that the code could use some refactoring in general. I've checked out the PR locally, and it looks great. Thanks! There are some comments, but I'll fix those along with some more refactors.

saibotk commented:
Thank you very much! :rocket: 

That's fine, but still thank you for your review, so that I can learn more about improving my Haskell code :)

saibotk commented:
And thank you very much for the nix shell config :) 
Otherwise, it would have been much harder to get up and running, just to quickly add such a "simple" feature ;)

saibotk commented:
Oh @FPtje sorry, for the ping, but we might also want to add the new config to the README :sweat_smile:

FPtje commented:
I've refactored the `LintMessage` in a75b520, and made the promised changes.

I've also looked at the command line arguments. It looks like it's indeed _very_ difficult to add parameters there. That code is not great.

By not great I mean...
https://github.com/FPtje/GLuaFixer/blob/a75b52075363774a84901cf1fb85adbbab1c48af/src/GLuaFixer/LintMain.hs#L188-L190

This is not representative Haskell code. I've looked at [`optparse-applicative`](https://hackage.haskell.org/package/optparse-applicative), which is the go-to command line parsing tool of Haskell. The one big problem it has in this case is that it cannot replicate the current command line. It doesn't like subcommands starting with `--`, and the `--indentation=` parser is really hard to replicate as well.

I think the best option there is to create a backwards incompatible command line interface using `optparse-applicative`, and fall back to the old cli parser if the `optparse-applicative` one fails.

-------

Ah yes, the Readme. Good point!

FPtje commented:
Added to `README.md`! :+1:

Cherry commented:
This is so cool. Great work @saibotk!
